File: py/34.py
#!/usr/bin/python3

from util import fac, digits

def foo(n):
    return sum(map(fac, digits(n)))

def check(n):
    """Return True if n is a curious number.

    >>> check(145)
    True

    """

    return n == foo(n)

for i in range(10, 100000):
    if check(i):
        print(i)
print(sum([i for i in range(10, 100000) if check(i)]))

File: hs/34.hs
module Main where

import Data.Array

fac = (listArray (0,9) (scanl (*) 1 [1..9]) !)

limit = until (\x -> x * fac 9 < 10 ^ x) succ 1

check a = do
  let b = sum [fac (fromEnum x - 48) | x <- show a]
  if a == b
    then return a
    else fail "Not a valid number"

search = do
  a <- [3..10 ^ limit]
  check a
  return a

main = do
  print $ sum search
